Added bit masks to remove unused FM ports in MC

bryates opened the pull request:
Addressing Issue #120

aehart commented:
@bryates Nice! Seems to work as expected. I have just a few suggestions:
- We want to avoid hard-coding the mask for a specific instance of the MC in the MatchCalculator template function, which you do here:
https://github.com/cms-tracklet/firmware-hls/blob/ead0f05406d0fead5a699dccb9f3e3847b445b11/TrackletAlgorithm/MatchCalculator.h#L332-L340
Instead, these should be determined from the LAYER, DISK, and PHISEC parameters. One way of doing this is by passing these to a `constexpr` function that returns the appropriate bit mask, e.g., for the TC:
https://github.com/cms-tracklet/firmware-hls/blob/ead0f05406d0fead5a699dccb9f3e3847b445b11/TrackletAlgorithm/TrackletCalculator_parameters.h#L40-L42
- Right now, you represent the bit mask using eight constants, but since each constant has a unique shift, you could just OR them all into a single mask. This is partially a matter of style, but having a single mask would probably make addressing the previous suggestion easier.
- This is a really minor thing, but the right-shift by `projseed` isn't really necessary on these lines:
https://github.com/cms-tracklet/firmware-hls/blob/ead0f05406d0fead5a699dccb9f3e3847b445b11/TrackletAlgorithm/MatchCalculator.h#L829-L871

pwittich commented on TrackletAlgorithm/MatchCalculator.h:
```diff
@@ -798,36 +826,52 @@ void MatchCalculator(BXType bx,
     if(newtracklet && goodmatch==true) { // Write out only the best match, based on the seeding 
       switch (projseed) {
       case 0:
+      if((FML1L2 & mask_L1L2) >> projseed) { //projseed 0-7 equivalent to shift
```
In this `switch` statement since the case label is the same as the value of `projseed` if you make the amount you shift by the same as the case label it will be more clear to the compiler that these are constant expressions (assuming the shift doesn't go away like @aehart says)

pwittich replied:
also probably the switch statement argument should not be a magic number.

pwittich commented on TrackletAlgorithm/MatchCalculator.h:
```diff
@@ -311,6 +311,34 @@ void MatchCalculator(BXType bx,
+ 
+  // MC_L3PHIC mask {1: on, 0: off}
+  static const uint16_t FML1L2 = 1 << shift_L1L2;
+  static const uint16_t FML2L3 = 0 << shift_L2L3;
```
what does it mean to shift a zero? I am confused. And are these static const or constexpr?

bryates commented:
@aehart thanks for the pointers. I've changed things over to look more similar to the tracklet calculator. The mask is now stored in a constexpr in MatchCalculator_parameters.h, and projseed was completely removed from the mask evaluation. This also adresses the comment from @pwittich about projseed.

pwittich commented on TrackletAlgorithm/MatchCalculator.h:
```diff
@@ -314,30 +319,30 @@ void MatchCalculator(BXType bx,
   static const uint8_t shift_L1L2 = 0;
```
these are still static const -- can't they be constexpr?

bryates replied:
Moved to MatchCalculator_parameters.h

pwittich commented on TrackletAlgorithm/MatchCalculator.h:
```diff
@@ -826,49 +831,49 @@ void MatchCalculator(BXType bx,
       switch (projseed) {
       case 0:
-      if((FML1L2 & mask_L1L2) >> projseed) { //projseed 0-7 equivalent to shift
+      if(FMMask<LAYER, PHISEC>() & mask_L1L2) { //projseed 0-7 equivalent to shift
```
lots of bad formatting here please fix

bryates replied:
Cleaned up

pwittich commented on TestBenches/MatchCalculator_test.cpp:
```diff
@@ -75,7 +75,7 @@ int main() {
   // loop over events
   for (int ievt = 0; ievt < nevents; ++ievt) {
-    //cout << "Event: " << dec << ievt << endl;
+    cout << "Event: " << dec << ievt << endl;
```
probably should wrap this into an `ifdef __SYNTHESIS__`

bryates replied:
I forgot to comment that line out again after testing. It is now commented.

bryates commented:
I've cleaned up the code a bit, and moved the shift and mask values to MatchCalculator_parameters.h. They should not have to bee constexpr, but @pwittich please let me know if you'd prefer they are.

pwittich commented:
There are a lot of magic numbers in this code (What do you think @aehart @tomalin ) Can you get rid of some of them with `auto` and otherwise get rid of them with meaningful constants?

aehart reviewed (commented):
> There are a lot of magic numbers in this code (What do you think @aehart @tomalin ) Can you get rid of some of them with `auto` and otherwise get rid of them with meaningful constants?

I agree. Liberal use of `auto` in particular will go a long way toward cleaning up the magic bit widths.

bryates commented on TrackletAlgorithm/MatchCalculator.h:
```diff
@@ -145,35 +145,35 @@ ap_uint<width> iabs( ap_int<width> value )
 void readTable_Cuts(ap_uint<width> table[depth]){
   if (phi){ // phi cuts
-    if (L==1){
+    if (L==0){
```
Simple enough, I'll do that now.

bryates commented on TrackletAlgorithm/MatchCalculator.h:
```diff
@@ -225,8 +225,14 @@ void readTable_Cuts(ap_uint<width> table[depth]){
 
 // MatchCalculator
+namespace MC {
+  enum itc {UNDEF_ITC, A = 0, B = 1, C = 2, D = 3, E = 4, F = 5, G = 6, H = 7, I = 8, J = 9, K = 10, L = 11, M = 12, N = 13, O = 14};
```
Good point, I'll make that change.

bryates commented:
Several changes in the latest commit:
- itc -> imc
- shifts completely replaced with TF numbers 
https://github.com/cms-tracklet/firmware-hls/blob/aba7d82bf32bc440ee78482f285ef0b0af2871ed/TrackletAlgorithm/MatchCalculator_parameters.h#L7-L15
- table reads now use TF numbers
https://github.com/cms-tracklet/firmware-hls/blob/aba7d82bf32bc440ee78482f285ef0b0af2871ed/TrackletAlgorithm/MatchCalculator.h#L151

bryates commented:
One more change. I was able to completely remove the mask constants, and just use the seed values directly.
https://github.com/cms-tracklet/firmware-hls/blob/01c5f17e6748895fb8e156a0bc6f5488f5e16a86/TrackletAlgorithm/MatchCalculator_parameters.h#L12-L14

aehart reviewed (approved):
Looks good to me.
